### ISSUE TITLE
Request: show-smartparens and offscreen delimiter

### ISSUE THREAD

**Issue by elemakil** [Fuco1/smartparens#346]
On the [EmacsWiki](http://www.emacswiki.org/emacs/ShowParenMode#toc1) there is a nice `advice` to the builtin `show-paren-function`, which will print the matching delimiter in the echo area if it is offscreen. 

I'd like to put in a request for this feature for the `show-smartparens-mode`. I have adapted the solution given in the EmacsWiki article for `smartparens` (see below code), however, this solution is limited to the _traditional_ paren pairs and also I'd assume that there is a cleverer solution than this.

``` elisp
;; show matching parens on steroids
(defadvice sp-show--pair-function
  (after sp-show--pair-function-offscreen activate)
  "If the matching paren is offscreen, show the matching line in the
        echo area. Has no effect if the character before point is not of
        the syntax class ')'."
  (interactive)
  (if (not (minibuffer-prompt))
      (let ((matching-text nil))
        ;; Only call `blink-matching-open' if the character before point
        ;; is a close parentheses type character. Otherwise, there's not
        ;; really any point, and `blink-matching-open' would just echo
        ;; "Mismatched parentheses", which gets really annoying.
        ;; (if (char-equal (char-syntax (char-before (point))) ?\))
        ;;     (setq matching-text (blink-matching-open)))
        ;; (if (not (null matching-text))
        ;;     (message matching-text)))))
        (let* ((cb (char-before (point)))
               (matching-text (and cb
                                   (char-equal (char-syntax cb) ?\) )
                                   (blink-matching-open))))
          (when matching-text (message matching-text))))))
```

**Comment by rgemulla** [comment]
Another option below. This code directly uses the matching-pair overlay of smartparen (and thus avoids expensive recomputation of the pair or the use of another library).

```lisp
  ;; also show the text of matching paren in echo area
  (require 'subr-x)
  (with-no-warnings
    (make-variable-buffer-local 'rg/sp-previous-overlay)
    (make-variable-buffer-local 'rg/sp-previous-point))
  (defun rg/sp-show-match-in-echo-area ()
    "If the point is on an opening or closing paren and
show-smartparens-mode is active, show the line of the matching
paren in the echo area if it's not visible in the window."
    (interactive)
    (when show-smartparens-mode
      (let* ((start (car sp-show-pair-overlays))
             (end (cdr sp-show-pair-overlays))
             (message-log-max)) ;; don't log in messages
        (if (and start end)
            (let ((current-overlay (list (overlay-start start)
                                         (overlay-end start)
                                         (overlay-start end)
                                         (overlay-end end))))
              (when (not (and (equal rg/sp-previous-overlay current-overlay)
                              (equal (point) rg/sp-previous-point)))
                (save-excursion
                  (let* ((where (if (<= (point) (overlay-end start))
                                    ;; opening delim is selected
                                    (overlay-end end)
                                  ;; closing delim is selected
                                  (overlay-start start)))
                         (visible (pos-visible-in-window-p where)))
                    (when (not visible)
                      (let* ((from (progn (goto-char where) (beginning-of-line) (point)))
                             (to (progn (end-of-line) (point)))
                             (s (buffer-substring from to)))
                        ;; Copy smartparens overlay for opening parens
                        (let ((i1 (- (nth 0 current-overlay) from))
                              (i2 (- (nth 1 current-overlay) from)))
                          (when (and (< i1 (length s)) (>= i2 0))
                            (add-face-text-property (max i1 0) (min i2 (length s))
                                                    'sp-show-pair-match-face nil s)))
                        ;; Copy smartparens overlay for closing parens
                        (let ((i1 (- (nth 2 current-overlay) from))
                              (i2 (- (nth 3 current-overlay) from)))
                          (when (and (< i1 (length s)) (>= i2 0))
                            (add-face-text-property (max i1 0) (min i2 (length s))
                                                    'sp-show-pair-match-face nil s)))
                        ;; echo line of match
                        (message "Matches: %s" (string-trim s)))))))
              (setq rg/sp-previous-overlay current-overlay)
              (setq rg/sp-previous-point (point)))
        (setq rg/sp-previous-overlay nil)))))
  (advice-add 'sp-show--pair-function :after 'rg/sp-show-match-in-echo-area)
```

**Comment by Fuco1** [comment]
This seems quite complicated.  Wouldn't it be enough to grab the overlay, check if the point is visible and echo the line?

Why do you need to store the previous overlay/point?

**Comment by rgemulla** [comment]
I'll keep it to display the message only once. This turned out to be necessary. For example, when evaluating a form in Lisp mode via `C-x C-e` (or, in fact, any command), the result will be printed in the echo area. If the previous overlay/point were not stored, this message would be overwritten immediately by the echoed matching paren (which is annoying at best).

The other more complicated part copies the overlay into the echoed message so that one can see which paren is matched, which is useful whenever there is more than one.

**Comment by Fuco1** [comment]
I see, yes, that sounds reasonable.

I think it would be best if we integrated this directly into the matching function in SP instead of as an advice.  There should be some feature flag of course that would toggle the behaviour.

Do you think you could make such a patch? That would be great!

**Comment by rgemulla** [comment]
I agree. Pull request #787.

**Comment by rgemulla** [comment]
I created new pull request #788, which only has the commits related to this issue.

**Comment by Fuco1** [comment]
Closed by #788.  Thanks everyone for ideas/code!
